crates/rocksdb/src/db_manager.rs

```diff
@@ -0,0 +1,330 @@
+                    watchdog.on_config_update();
+                }
+            }
+        }
+
+        watchdog.shutdown();
+        Ok(())
+    }
+
+    fn shutdown(&mut self) {
+        // Ask all databases to shutdown cleanly.
+        let start = Instant::now();
+        for v in &self.manager.dbs {
+            info!(
+                db = %v.name,
+                owner = %v.owner,
+                "Shutting down rocksdb database"
+            );
+            if let Err(e) = v.db.flush_wal(true) {
+                warn!(
+                    db = %v.name,
+                    owner = %v.owner,
+                    "Failed to flush local loglet rocksdb WAL: {}",
+                    e
+                );
+            }
+            if let Err(e) = v.db.flush_memtables_atomically(v.cfs().as_slice(), true) {
+                warn!(
+                    db = %v.name,
+                    owner = %v.owner,
+                    "Failed to flush memtables: {}",
+                    e
+                );
+            }
+            v.db.cancel_all_background_work(true);
+        }
+        info!("Clean rocksdb shutdown took {:?}", start.elapsed());
+    }
+
+    fn handle_command(&mut self, cmd: WatchdogCommand) {
+        match cmd {
+            WatchdogCommand::ResetAll => {
+                self.shutdown();
+                self.manager.dbs.clear();
+                self.subscriptions.clear();
+            }
+            WatchdogCommand::Register(sub) => self.subscriptions.push(sub),
+        }
+    }
+
+    fn on_config_update(&mut self) {
+        // Memory budget changed?
+        let new_common_opts = self.updateable_common_opts.load();
+        if new_common_opts.rocksdb_total_memory_limit
+            != self.current_common_opts.rocksdb_total_memory_limit
+        {
+            info!(
+                old = self.current_common_opts.rocksdb_total_memory_limit,
+                new = new_common_opts.rocksdb_total_memory_limit,
+                "[config update] Setting rocksdb total memory limit to {}",
+                ByteSize::b(new_common_opts.rocksdb_total_memory_limit)
+            );
+            self.cache
+                .set_capacity(new_common_opts.rocksdb_total_memory_limit as usize);
+        }
+
+        // update memtable total memory
+        if new_common_opts.rocksdb_total_memtables_size_limit
+            != self.current_common_opts.rocksdb_total_memtables_size_limit
+        {
+            info!(
+                old = self.current_common_opts.rocksdb_total_memtables_size_limit,
+                new = new_common_opts.rocksdb_total_memtables_size_limit,
+                "[config update] Setting rocksdb total memtables size limit to {}",
+                ByteSize::b(new_common_opts.rocksdb_total_memtables_size_limit)
+            );
+            self.manager
+                .write_buffer_manager
+                .set_buffer_size(new_common_opts.rocksdb_total_memtables_size_limit as usize);
```

Comment:
What happens if the current usage exceeds the new limit? Is the limit a soft limit and RocksDB would simply use more than its limit and resize the write buffers when having flushed the old ones?

Reply:
Yes, all those limits are soft limits, maybe I should make the config be named to reflect this.